core/src/main/java/jenkins/install/UpgradeWizard.java

```diff
@@ -0,0 +1,147 @@
+package jenkins.install;
+
+import hudson.Extension;
+import hudson.model.PageDecorator;
+import hudson.model.UpdateSite.Plugin;
+import hudson.util.HttpResponses;
+import hudson.util.VersionNumber;
+import jenkins.model.Jenkins;
+import org.apache.commons.io.FileUtils;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.HttpResponse;
+import org.kohsuke.stapler.interceptor.RequirePOST;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+import static java.util.logging.Level.*;
+import static org.apache.commons.io.FileUtils.*;
+import static org.apache.commons.lang.StringUtils.*;
+
+/**
+ * This is a stop-gap measure until JENKINS-33663 comes in.
+ * This call may go away. Please don't use it from outside.
+ *
+ * @author Kohsuke Kawaguchi
+ */
+@Restricted(NoExternalUse.class)
+@Extension
+public class UpgradeWizard extends PageDecorator {
+    @Inject
+    private Jenkins jenkins;
+
+    /**
+     * Is this instance fully upgraded?
+     */
+    private volatile boolean upToDate;
+
+    /**
+     * File that captures the state of upgrade.
+     *
+     * This file records the vesrion number that the installation has upgraded to.
+     */
+    /*package*/ File getStateFile() {
+        return new File(Jenkins.getInstance().getRootDir(),"upgraded");
```

Comment:
I looked into this, and the upgrade/downgrade handling scheme in those classes are not readily usable here. The problem boils down to when the transition between versions are considered complete.

The `InstallUtil` version number is used to perform detached plugin installations, and this happens during boot without user interactions. In contrast, here we are trying to do something with user consent and possibly much later after upgrade has actually happened.